plutus-chain-index-core/src/Plutus/ChainIndex/Emulator/Handlers.hs

```diff
@@ -123,33 +105,51 @@ getUtxoutFromRef ::
   , Member (LogMsg ChainIndexLog) effs
   )
   => TxOutRef
-  -> Eff effs (Maybe L.ChainIndexTxOut)
+  -> Eff effs (Maybe L.OffChainTxOut)
 getUtxoutFromRef ref@TxOutRef{txOutRefId, txOutRefIdx} = do
   ds <- gets (view diskState)
   -- Find the output in the tx matching the output ref
   case preview (txMap . ix txOutRefId . citxOutputs . _ValidTx . ix (fromIntegral txOutRefIdx)) ds of
-    Nothing -> logWarn (TxOutNotFound ref) >> pure Nothing
-    Just txout@(ChainIndexTxOut address value datum refScript) -> do
-      -- The output might come from a public key address or a script address.
-      -- We need to handle them differently.
-      case addressCredential $ citoAddress txout of
-        PubKeyCredential _ ->
-          pure $ Just $ L.PublicKeyChainIndexTxOut address value datum script
-        ScriptCredential vh@(ValidatorHash h) -> do
-          case datum of
-            OutputDatumHash dh -> do
-              let v = maybe (Left vh) (Right . Validator) $ preview (scriptMap . ix (ScriptHash h)) ds
-              let d = maybe (Left dh) Right $ preview (dataMap . ix dh) ds
-              pure $ Just $ L.ScriptChainIndexTxOut address value d script v
-            OutputDatum d -> do
-              let v = maybe (Left vh) (Right . Validator) $ preview (scriptMap . ix (ScriptHash h)) ds
-              pure $ Just $ L.ScriptChainIndexTxOut address value (Right d) script v
-            NoOutputDatum -> do
-              -- If the txout comes from a script address, the Datum should not be Nothing
-              logWarn $ NoDatumScriptAddr txout
-              pure Nothing
-      where
-        script = fromReferenceScript refScript
+    Nothing -> do
+      logWarn (TxOutNotFound ref)
+      pure Nothing
+    Just txOut ->
+      makeOffChainTxOut txOut
+
+makeOffChainTxOut ::
+  forall effs.
+  ( Member (State ChainIndexEmulatorState) effs
+  , Member (LogMsg ChainIndexLog) effs
+  )
+  => ChainIndexTxOut
+  -> Eff effs (Maybe L.OffChainTxOut)
+makeOffChainTxOut txOut@(ChainIndexTxOut address value datum refScript) = do
+  ds <- gets (view diskState)
+  -- The output might come from a public key address or a script address.
+  -- We need to handle them differently.
+  case addressCredential address of
+    PubKeyCredential pkh ->
+      case datum of
+        OutputDatumHash dh ->
+          pure $ Just $ L.mkPublicKeyOffChainTxOut pkh value (Just (Left dh)) (fromReferenceScript refScript)
```

Comment:
Why not change `mkPublicKeyOffChainTxOut` to use the `OutputDatum` type?

Reply:
Good question, `OutputDatum` is only present in the V2 `plutus-ledger-api` and I wanted to avoid having to expose V2 stuff to all code consuming `plutus-ledger` (in this particular case it's chain-index that decided to use `OutputDatum`, other representations were possible).

Reply:
Another option would be to have a `OutputDatum -> Maybe (Either DatumHash Datum)` function, to avoid too much repetition here. (The rest seems to be the same, but I might overlook something)

Reply:
@sjoerdvisscher I agree. I feel `plutus-ledger-api` could offer more functions to use `OutputDatum` but nobody prevents us from introducing an internal helper here to reduce duplication. :+1: